### PR TITLE
[DTensor] fix copy_ strategy

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158538
* #158490

The previous strategy directly used 'self' input strategy for 'src'
input.  The fixed strategy correctly maps the self dim to src dim
so that it works even if the src input is broadcast.

E.g. for this program, broadcasting will occur on dims 0,1,3 of self.

```
self = torch.ones((2,3,4,5))
src = torch.ones((4,1))
self.copy_(src)
```

These are the correct sharding combinations:

|   self   |     src |
|-------|------|
| Shard(0)  |   Replicate() |
| Shard(1)  |   Replicate() |
| Shard(2)  |   Shard(0) |
| Shard(3)  |   Shard(1) |

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @pragupta

### CODE REVIEW

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 1]
```diff
@@ -93,44 +93,54 @@ def propagate_single_input_strategy(op_schema: OpSchema) -> StrategyType:
...
+                        mesh=mesh,
+                        placements=tuple(mapped_src_placements),
+                        tensor_meta=src_meta,
+                    ),
                 ],
             )
-            for strategy in first_input_strategy.strategies
-        ]
-    )
+        )
```
Comment (zpcore) [inline torch/distributed/tensor/_ops/_tensor_ops.py]: The specs looks good to me, can you also add the redistribute_cost?

Comment (wconstab) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: oh yea! thanks. i had it in my first version and then forgot about it.

Can we actually make it a required argument for creating an OpSpec?  (Unless we do your proposal of autogenerating them always, which i prefer)

Comment (zpcore) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: Yea, I think we can make it a required field once we fix all currently supported ops in the list https://github.com/pytorch/pytorch/issues/157495.

**Review by zpcore — APPROVED** [review]
LGTM!

File: test/distributed/tensor/test_tensor_ops.py [hunk 2]
```diff
@@ -68,22 +68,37 @@ def test_copy_(self):
...
-    #     for dst_spec, src_spec in zip(dst_specs, src_specs):
-    #         src_dtensor = distribute_tensor(src_tensor, device_mesh, dst_spec)
-    #         dst_dtensor = distribute_tensor(dst_tensor, device_mesh, src_spec)
-    #         # perform a broadcasted copy from Shard(0) to Shard(1) for the worst case
-    #         dst_dtensor.copy_(src_dtensor)
-    #         dst_tensor.copy_(src_tensor)
-    #         self.assertEqual(dst_dtensor.full_tensor(), dst_tensor)
+    @with_comms
+    def test_copy_broadcast_redistribute(self):
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
```
Comment (XilunWu) [inline test/distributed/tensor/test_tensor_ops.py]: i suggest we universally use the `init_device_mesh` API instead

File: test/distributed/tensor/test_tensor_ops.py [hunk 3]
```diff
@@ -68,22 +68,37 @@ def test_copy_(self):
...
+        # The src specs in this case are designed to not be compatible with the dst_specs, redistribute should happen
+        src_specs = [[Shard(1)], [Shard(1)], [Shard(1)]]
```
Comment (XilunWu) [inline test/distributed/tensor/test_tensor_ops.py]: `src_spec = [Shard(1)]`. Remove `src_specs` from the loop variable.

File: test/distributed/tensor/test_tensor_ops.py [hunk 4]
```diff
@@ -68,22 +68,37 @@ def test_copy_(self):
...
+        src_tensor = torch.randn((64, 1))
+
+        dst_tensor = torch.zeros(16, 32, 64, 128)
+        dst_specs = [[Replicate()], [Shard(1)], [Shard(2)]]
```
Comment (XilunWu) [inline test/distributed/tensor/test_tensor_ops.py]: no need for [Shard(0)]?

File: test/distributed/tensor/test_tensor_ops.py [hunk 5]
```diff
@@ -68,22 +68,37 @@ def test_copy_(self):
...
+        for dst_spec, src_spec in zip(dst_specs, src_specs):
+            src_dtensor = distribute_tensor(src_tensor, device_mesh, src_spec)
+            dst_dtensor = distribute_tensor(dst_tensor, device_mesh, dst_spec)
+            dst_dtensor.copy_(src_dtensor)
+            dst_tensor.copy_(src_tensor)
+            self.assertEqual(dst_dtensor.full_tensor(), dst_tensor)
```
Comment (XilunWu) [inline test/distributed/tensor/test_tensor_ops.py:82]: since we're not tracking `comm` counts, use the `_run_test_on_dtensor` util?

File: test/distributed/tensor/test_tensor_ops.py [hunk 6]
```diff
@@ -68,22 +68,37 @@ def test_copy_(self):
...
+
+    @with_comms
+    def test_copy_broadcast(self):
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
```
Comment (XilunWu) [inline test/distributed/tensor/test_tensor_ops.py]: ditto

File: test/distributed/tensor/test_tensor_ops.py [hunk 7]
```diff
@@ -68,22 +68,37 @@ def test_copy_(self):
...
+        dst_specs = [[Replicate()], [Shard(1)], [Shard(2)]]
+        for dst_spec, src_spec in zip(dst_specs, src_specs):
+            src_dtensor = distribute_tensor(src_tensor, device_mesh, src_spec)
+            dst_dtensor = distribute_tensor(dst_tensor, device_mesh, dst_spec)
+            dst_dtensor.copy_(src_dtensor)
+            dst_tensor.copy_(src_tensor)
+            self.assertEqual(dst_dtensor.full_tensor(), dst_tensor)
+
+    @with_comms
+    def test_copy_broadcast(self):
```
Comment (XilunWu) [inline test/distributed/tensor/test_tensor_ops.py]: Do we need 2 separate tests for `copy_` + broadcast? Can they merge?

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 8]
```diff
@@ -93,44 +93,59 @@ def propagate_single_input_strategy(op_schema: OpSchema) -> StrategyType:
...
-    # TODO: this strategy is incorrect for copy_ in the case that src tensor
-    # is smaller rank than self tensor.  It is possible to select a strategy from self tensor
-    # that is invalid for dst tensor.
-    # It is also problematic to assume that shard(0) on src maps to shard(0) on self, since we
-    # may broadcast a new dim to the left or right of 0 when copying.
+def copy_inplace_strategy(op_schema: OpSchema) -> StrategyType:
+    assert isinstance(op_schema.args_schema[0], OpStrategy)
+    assert isinstance(op_schema.args_schema[1], OpStrategy)
+    self_strategy: OpStrategy = op_schema.args_schema[0]
+    mesh = self_strategy.mesh
```
Comment (XilunWu) [inline torch/distributed/tensor/_ops/_tensor_ops.py]: TODO: consider supporting cross-mesh copy

**Review by XilunWu — APPROVED** [review]
some nits

File: test/distributed/tensor/test_tensor_ops.py (hunk 4 above)
Comment (wconstab) [reply test/distributed/tensor/test_tensor_ops.py]: not really, its the same edge case as Shard(1), both of them are going to trigger the Shard->Redistribute path.  But i can add it, just for completeness.

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 8 above)
Comment (wconstab) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: hmm, is this possible?
at least, the meshes would have to have compatible shapes.  

if we already have this support in other places, then i should follow it here. Is it documented somewhere / example code to point to?

File: test/distributed/tensor/test_tensor_ops.py (hunk 2 above)
Comment (wconstab) [reply test/distributed/tensor/test_tensor_ops.py]: i am just sticking with the convention in the file. I also want to change things to MultiProcContinuousTest.  (or  threaded, via PR #158082, if that can be made to work - so ill leave it alone in this PR

File: test/distributed/tensor/test_tensor_ops.py (hunk 5 above)
Comment (wconstab) [reply test/distributed/tensor/test_tensor_ops.py:82]: actually, i should check the comm counts, i was just lazy.  I should ensure that a redistribute happend.

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 9]
```diff
@@ -93,44 +93,59 @@ def propagate_single_input_strategy(op_schema: OpSchema) -> StrategyType:
 
 
 @register_op_strategy(aten.copy_.default)
-def copy_strategy(op_schema: OpSchema) -> StrategyType:
-    # TODO: this strategy is incorrect for copy_ in the case that src tensor
-    # is smaller rank than self tensor.  It is possible to select a strategy from self tensor
-    # that is invalid for dst tensor.
-    # It is also problematic to assume that shard(0) on src maps to shard(0) on self, since we
-    # may broadcast a new dim to the left or right of 0 when copying.
+def copy_inplace_strategy(op_schema: OpSchema) -> StrategyType:
```
Comment (wanchaol) [inline torch/distributed/tensor/_ops/_tensor_ops.py]: Hmmm I think you probably want to remove the copy strategy, just register it as a pointwise strategy.

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 10]
```diff
@@ -93,44 +93,59 @@ def propagate_single_input_strategy(op_schema: OpSchema) -> StrategyType:
...
+    # e.g. the last dim of 'src' must match up with the last dim of 'self'
     #
-    # For now, I just keep copy working essentially the way it was before this PR,
-    # but split it out so it can be handled separately in the future.
-    num_tensor_args = 2
-    first_input_strategy = op_schema.args_schema[0]
-    assert isinstance(first_input_strategy, OpStrategy)
-    return OpStrategy(
-        [
+    # DTensor semantics for inplace ops also dictates that we may NOT redistribute our 'self' input.
```
Comment (wanchaol) [inline torch/distributed/tensor/_ops/_tensor_ops.py]: I don't quite like the fact that we are implementing broadcasting semantic separately, it would be better if you could just reuse the pointwise strategy for `copy_`

**Review by wanchaol — COMMENTED** [review]
Please see inlined comments, I think we should just use register copy_as a pointwise strategy in `_tensor_ops.py`

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 9 above)
Comment (wconstab) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: hmm. i would be ok with that, but does pointwise strategy properly enforce the extra requirement of inplace_ ops, where broadcasting 'self' to match 'src' is NOT allowed?

Comment (wanchaol) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: Yeah, for inplace ops it just follow the first argument

Comment (wconstab) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: ok, yea i think we can try pointwise. Wish I had known about it!

File: test/distributed/tensor/test_tensor_ops.py (hunk 3 above)
Comment (wconstab) [reply test/distributed/tensor/test_tensor_ops.py]: i don't understand? seems ok to me.

File: test/distributed/tensor/test_tensor_ops.py (hunk 7 above)
Comment (wconstab) [reply test/distributed/tensor/test_tensor_ops.py]: merged them

Comment (wconstab) [reply test/distributed/tensor/test_tensor_ops.py]: note- i prefer separate tests (or using parameterize) so its easy to run a subtest without commenting stuff out.

but its horribly slow with the current multiproc test case, so i merged them for now

**Review by wanchaol — APPROVED** [review]

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 8 above)
Comment (XilunWu) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: I see people are supporting cross-mesh ops to some degree such as in: https://github.com/pytorch/pytorch/pull/157682 and https://github.com/pytorch/pytorch/pull/157049

File: test/distributed/tensor/test_tensor_ops.py (hunk 2 above)
Comment (Electron4444) [reply test/distributed/tensor/test_tensor_ops.py]: I closed PR #158082 since I don't have the experience with the codebase yet, to know how to get it to work
so you can do those changes now if you want to

Propably better to split up those changes into seperate commits anyway
